[10.0][ADD] sale_blanket_order

andreparames opened the pull request:
This module adds blanket orders, as described in the corresponding issue.

This PR depends on another for the migration of `web_action_conditionable`: https://github.com/OCA/web/pull/845

Fixes #592

andreparames committed:
[10.0][ADD] sale_blanket_orders

pedrobaeza commented:
Please avoid plurals on module name.

Cedric-Pigeon commented on sale_blanket_orders/__manifest__.py:
```diff
@@ -0,0 +1,27 @@
+    'category': 'Sale',
+    'license': 'AGPL-3',
+    'author': 'Acsone,Odoo Community Association (OCA)',
+    'version': '11.0.1.0.0',
```
Version should be 10

Cedric-Pigeon commented on sale_blanket_orders/__manifest__.py:
```diff
@@ -0,0 +1,27 @@
+    'name': 'Blanket Orders',
+    'category': 'Sale',
+    'license': 'AGPL-3',
+    'author': 'Acsone,Odoo Community Association (OCA)',
```
Acsone SA/NV

Cedric-Pigeon commented on sale_blanket_orders/models/blanket_orders.py:
```diff
@@ -0,0 +1,232 @@
+        string='Ordered quantity', compute='_compute_quantities', store=True)
+    invoiced_qty = fields.Float(
+        string='Invoiced quantity', compute='_compute_quantities', store=True)
+    remaining_qty = fields.Float(
```
Perhaps you should add the delivered quantities also

Cedric-Pigeon commented on sale_blanket_orders/views/blanket_orders.xml:
```diff
@@ -0,0 +1,121 @@
+                            <field name="partner_id" domain="[('customer','=',True)]" context="{'search_default_customer':1, 'show_address': 1}" options='{"always_reload": True}'/>
+                        </group>
+                        <group>
+                            <field name="validity_date"/>
```
should be set required in draft state through an attrs attribute

andreparames replied:
No problem. I had added it initially, but then thought it could be useful to create the order before the validity date is agreed, so instead I added a check when the user confirms it.
I'll add them back.

Cedric-Pigeon commented on sale_blanket_orders/views/blanket_orders.xml:
```diff
@@ -0,0 +1,121 @@
+                        <page string="Other Information">
+                            <group>
+                                <group string="Sales Information" name="sales_person">
+                                    <field name="pricelist_id"/>
```
should be set required in draft state

atchuthan commented on sale_blanket_order/data/ir_cron.xml:
```diff
@@ -0,0 +1,20 @@
+<!-- Copyright 2018 ACSONE SA/NV
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). -->
+
+<odoo noupdate="0">
```
should be noupdate="1"

atchuthan reviewed (approved):
Code review :+1:

andreparames commented:
Thanks @pedrobaeza @Cedric-Pigeon @atchuthan for the review. The checks are failing because the module depends on `web_action_conditionable`, which I've migrated and is also waiting to be merged: https://github.com/OCA/web/pull/845

Commit:
[ADD] add setup for sale_blanket_order

Cedric-Pigeon commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -0,0 +1,235 @@
+    )
+    def _compute_quantities(self):
+        for line in self:
+            sale_lines = self.sale_order_lines_ids
```
you should use `line.` instead of `self`

Cedric-Pigeon reviewed (approved):
Code review and functionnal tests

misern2 reviewed (changes requested):
Thanks for the contribution.
Just a minor change:
 if you duplicate the blanket order it keeps the same sequence and state.

andreparames committed:
sale_blanket_order: don't copy name nor confirm state

When duplicating a confirmed blanket order, the new copy shouldn't
keep the state nor the sequence number (name).

andreparames commented:
Thanks @misern2, I've pushed a commit that fixes that problem, and also lets the lines be copied when duplicating an order.

misern2 reviewed (approved):
:+1: LGTM . Thanks!.

andhit-r commented:
@andreparames thanks for contribution. Few things to consider:

Sequence
-------------

Since sequence will be assigned when blanket order confirmed, it will leave draft BO with **draft** identifier. IMO, it will give user a problem to identify draft BO.

**Propose Solution**

1. If you still want to assign sequence when BO confirmed, make display_name *{ID} like in account.move. Or,
2. Assign sequence when BO saved. With default name value == "/".

State
-------

It always useful to have approval step after confirm

**Propose Solution**

Draft -> Waiting for Approval -> Opened

If you agree my points above, i will create PR for those propose solutions.

Many thanks.

rousseldenis commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -0,0 +1,329 @@
+    payment_term_id = fields.Many2one(
+        'account.payment.term', string='Payment Terms', readonly=True,
+        states={'draft': [('readonly', False)]})
+    confirmed = fields.Boolean(default=False)
```
That's default

andreparames replied:
"Explicit is better than implicit" :smile:

rousseldenis commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -0,0 +1,329 @@
+    def _get_sale_orders(self):
+        lines = self.mapped('lines_ids')
+        sale_lines = lines.mapped('sale_order_lines_ids')
+        sale_orders = sale_lines.mapped('order_id')
```
You can fetch them through one mapped('line_ids.sale_order_lines_ids.order_id')

rousseldenis commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -0,0 +1,329 @@
+        sale_lines = lines.mapped('sale_order_lines_ids')
+        sale_orders = sale_lines.mapped('order_id')
+        sale_orders_ids = list(set(sale_orders.ids))
+        return self.env['sale.order'].browse(sale_orders_ids)
```
Return just the result of mapped.

rousseldenis commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -0,0 +1,329 @@
+        return self.env['sale.order'].browse(sale_orders_ids)
+
+    @api.multi
+    @api.depends('lines_ids.remaining_qty')
```
Is that correct ?

rousseldenis commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -0,0 +1,329 @@
+                assert len(order.lines_ids) > 0, _("Must have some lines")
+                order.lines_ids._validate()
+        except AssertionError as e:
+            raise Warning(e.message)
```
Deprecated, use UserError

rousseldenis commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -0,0 +1,329 @@
+    confirmed = fields.Boolean(default=False)
+    state = fields.Selection(selection=[
+        ('draft', 'Draft'),
+        ('opened', 'Opened'),
```
Maybe 'open' is more correct

rousseldenis commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -0,0 +1,329 @@
+    sale_order_lines_ids = fields.One2many(
+        'sale.order.line', 'blanket_line_id', string='Sale order lines')
+    company_id = fields.Many2one(
+        'res.company', related='order_id.company_id', store=True)
```
Readonly=True is recommended for related fields

rousseldenis commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -0,0 +1,329 @@
+                assert line.original_qty > 0.0, \
+                    _("Quantity must be greater than zero")
+        except AssertionError as e:
+            raise Warning(e.message)
```
Deprecated

rousseldenis commented on sale_blanket_order/models/sale_orders.py:
```diff
@@ -0,0 +1,18 @@
+
+    blanket_order_id = fields.Many2one(
+        'sale.blanket.order', string='Origin blanket order',
+        related='order_line.blanket_line_id.order_id')
```
readonly=True

rousseldenis commented on sale_blanket_order/wizard/create_sale_orders.py:
```diff
@@ -0,0 +1,111 @@
+                continue
+
+            if line.qty > line.remaining_qty:
+                raise Warning(
```
Deprecated

rousseldenis reviewed (changes requested):
Some comments.

One remark : why don't you use decimal precision on quantities ?

andreparames committed:
sale_blanket_order: update after review

andreparames commented:
Thanks for the review @rousseldenis!

andreparames commented:
@andhit-r they seem like good suggestions. I'd just add an option to enable the approval step.

heathfx commented:
I stumbled across this module while trolling the web and hope I'm not intruding by commenting here. This module could be very useful at my company and when I was testing it, noticed that the price compute method is likely missing an onchange decorator for when qty is updated.

This is important when pricelists are in use and different price tiers are based on qty.

andreparames committed:
sale_blanket_order: Re-calculate price based on qty

andreparames commented:
@heathfx good point, I've made the onchange recalculate the price when the initial quantity changes.

andreparames committed:
sale_blanket_order: make prohibition of adding new lines optional

andreparames commented:
I've pushed an update that adds a configuration option to allow or disallow adding new lines to SOs created from BOs (by default it allows).

rousseldenis committed:
[FIX] sale_blanket_order: Ondelete cascade on blanket lines

AdriaGForgeFlow commented on sale_blanket_order/views/blanket_orders.xml:
```diff
@@ -0,0 +1,122 @@
+                    </div>
+                    <group>
+                        <group>
+                            <field name="partner_id" domain="[('customer','=',True)]" context="{'search_default_customer':1, 'show_address': 1}" options='{"always_reload": True}'/>
```
I think it should be set to required in draft state to make it clear for the user. In the _validate (blanket_orders.py) function you specify that Partner field is mandatory but that field does not appear in purple color in the web interface and that's a little bit confusing IMO.

rousseldenis replied:
@Cedric-Pigeon What do you think of this ?

Cedric-Pigeon replied:
it makes sense IMO

bjeficent commented:
Just a question:
If you Create a blanket order > Create sale order and instead of confirm it you cancel it and then remove it, the ordered quantity of that sale order and the "updated" remaining quantity is shown. Is this how is it supposed to work?

JordiBForgeFlow commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -0,0 +1,331 @@
+        'sale_order_lines_ids.qty_invoiced',
+        'original_qty'
+    )
+    def _compute_quantities(self):
```
I think that you need to add api.depends('sale_order_lines_ids', 'sale_order_lines_ids.ordered_qty',  'sale_order_lines_ids.invoiced_qty', 'sale_order_lines_ids.delivered_qty'. Otherwise there will always be discrepancies.

bjeficent replied:
You are right, those lines need to be added but without the sale_order_lines_ids in ordered_qty, invoiced_qty... I mean:
api.depends('sale_order_lines_ids', 'ordered_qty', 'invoiced_qty', 'delivered_qty', 'remaining_qty')

JordiBForgeFlow commented:
@bjeficent You are right. See my comment. Perhaps you can propose the fix.

nikul-serpentcs commented on sale_blanket_order/README.rst:
```diff
@@ -0,0 +1,69 @@
+
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/167/11.0
```
@andreparames `target` 10.0 :smiley:

nikul-serpentcs commented on sale_blanket_order/__manifest__.py:
```diff
@@ -0,0 +1,29 @@
+        'sale',
+        'web_action_conditionable',
+    ],
+    'data': [
```
@andreparames Sequence should be [this](https://github.com/OCA/maintainer-tools/blob/master/template/module/__manifest__.py#L29)

bjeficent commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -0,0 +1,331 @@
+    remaining_qty = fields.Float(
+        string='Remaining quantity', compute='_compute_quantities', store=True)
+    delivered_qty = fields.Float(
+        string='Delivered quantity', compute='_compute_quantities', store=True)
```
I think the store=True parameter should be removed from ordered_qty, invoiced_qty, remaining_qty and delivered_qty to fix that issue.

rousseldenis replied:
@bjeficent Why removing store=True ?

nikul-serpentcs reviewed (commented):
Improve Code

rgl1234 commented:
At the moment we see the following missing features for this functionality:

1.) possibility to have an analytic-account on BO if needed

2.) In our case, the products and quantities can change during the lifetime of a blanket order. So it would be good if the blanket-order could be modified after approving it and after having started the deliveries on a BO (add/remove products, change quantities on products, change validity date...)

does this maybe make sence too for someone else?

René

rgl1234 commented:
At the moment we see the following missing features for this functionality:

1.) possibility to have an analytic-account on BO if needed

2.) In our case, the products and quantities can change during the lifetime of a blanket order. So it would be good if the blanket-order could be modified after approving it and after having started the deliveries on a BO (add/remove products, change quantities on products, change validity date...)

does this maybe make sence too for someone else?

René

daramousk commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -0,0 +1,331 @@
+        for order in self:
+            if not order.confirmed:
+                order.state = 'draft'
+            elif order.validity_date <= today:
```
If the order's validity date is today, then the order is considered expired? Is the equals needed?

daramousk commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -0,0 +1,331 @@
+            return
+
+        values = {
+            'pricelist_id': (self.partner_id.property_product_pricelist and
```
Is it perhaps possible to just do `self.partner_id.property_product_pricelist.id` since this (despite it's name) is a `Many2one`? Same for the `property_payment_term_id`

daramousk commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -0,0 +1,331 @@
+        self.update(values)
+
+    @api.multi
+    def copy_data(self, default=None):
```
From what I understand, here you are getting the default values of the fields `name` and `confirmed` and injecting them on the newly created record.

Is this override necessary though? Because `create()` calls the `_add_missing_default_values()` which in turn calls `default_get` and adds the values later. So removing this would make sense. Or am I missing something here?

rousseldenis replied:
A simple copy=False will be sufficent

daramousk commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -0,0 +1,331 @@
+        return super(BlanketOrder, self).copy_data(default)
+
+    @api.multi
+    def _validate(self):
```
Maybe this, with it's assertions, could be replaced by `@api.constrains('confirmed')` and conditionals?

So it does not explicitly requires us to call it and it would be called when necessary.
Also, there are certain flags passed to the interpreter which can remove the assertions

rousseldenis replied:
See here under message

daramousk commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -0,0 +1,331 @@
+        sale_orders = self._get_sale_orders()
+        action = self.env.ref('sale.action_orders').read()[0]
+        if len(sale_orders) > 0:
+            action['domain'] = [('id', 'in', sale_orders.ids)]
```
The `action['domain']` already has a domain which is removed when this assignment takes place. Perhaps it would be better to append it?

daramousk commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -0,0 +1,331 @@
+    def action_view_sale_orders(self):
+        sale_orders = self._get_sale_orders()
+        action = self.env.ref('sale.action_orders').read()[0]
+        if len(sale_orders) > 0:
```
If you want to see whether your recordset has any records in it, you can avoid the `len()` altogether since `BaseModel.__nonzero__` is implemented. Therefore `if sale_orders` will suffice.

daramousk commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -0,0 +1,331 @@
+            line.remaining_qty = line.original_qty - line.ordered_qty
+
+    @api.multi
+    def _validate(self):
```
Same here, could this be perhaps replaced by `api.constrains()` and conditions instead of assert?

rousseldenis replied:
I'll modify asserts but this cannot be constrains because it's called on order validation (action)

rousseldenis replied:
And finally I think I will let asserts because I like the way it was done (catching AssertionError and raise an Odoo Exception)

daramousk commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -0,0 +1,331 @@
+        readonly=True)
+
+    def _get_real_price_currency(self, product, rule_id, qty, uom,
+                                 pricelist_id):
```
`pricelist_id` seems to not be used anywhere here.

rousseldenis replied:
Yes but can be overriden

daramousk commented on sale_blanket_order/models/blanket_orders.py:
```diff
@@ -0,0 +1,331 @@
+        """Retrieve the price before applying the pricelist
+            :param obj product: object of current product record
+            :parem float qty: total quentity of product
+            :param tuple price_and_rule: tuple(price, suitable_rule) coming
```
Perhaps `price_and_rule` should be renamed to `rule_id`?

daramousk reviewed (commented):
Hello and thank you for this, looks great.
I have left some comments mostly about the assertions.

Let me know what you think.

rafaelbn commented:
Hello @andreparames , please could you check bots are green :green_heart: ? I see travis red :red_circle:

rousseldenis commented:
Hi @rafaelbn, @andreparames does not wor k anymore on Odoo. I will try to finish this one.

rafaelbn commented:
Hi @rousseldenis , if you cannot work on this we can close this PR

rousseldenis commented:
@rafaelbn @daramousk Remarks that seems relevant to me has been attended. Waiting travis but ok for me.

Cedric-Pigeon commented:
@rafaelbn This one is ready to be merged.

rafaelbn commented:
/ocabot merge
